File: src/main/java/me/wuzzyxy/dynamicmarket/market/MarketManager.java
package me.wuzzyxy.dynamicmarket.market;

import me.wuzzyxy.dynamicmarket.DynamicMarket;
import me.wuzzyxy.dynamicmarket.database.Database;
import me.wuzzyxy.dynamicmarket.items.MarketItem;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class MarketManager {

    private final DynamicMarket plugin;
    private final Database database;

    private final MarketInitializer initializer;
    private final MarketDatabaseHandler databaseHandler;
    private final PriceHandler priceHandler;

    private final List<MarketItem> workingItems;
    private final List<MarketItem> persistedItems;
    public MarketManager(DynamicMarket plugin, Database database) {
        this.plugin = plugin;
        this.database = database;

        workingItems = new ArrayList<MarketItem>();
        persistedItems = getAllDBItems();
        if (persistedItems == null) {
            plugin.getLogger().severe("Failed to load items from database");
        }else {
            for (MarketItem item : persistedItems) {
                if (item == null) {
                    plugin.getLogger().severe("Failed to load items from database");
                    continue;
                }
                workingItems.add(item.clone());
            }
        }

        this.databaseHandler = new MarketDatabaseHandler(this, database, plugin);
        this.initializer = new MarketInitializer(plugin.getItemConfig(), this);
        this.priceHandler = new PriceHandler(plugin.getPluginConfig());

    }

    private List<MarketItem> getAllDBItems() {
        return database.getAllItems();
    }

    /***
     * Returns the working item object
     */
    public Optional<MarketItem> addItem(String name, double basePrice, double minPrice, double percentage) {
        if (getPersistedItem(name).isPresent()) return Optional.empty();

        MarketItem item = database.addItem(name, basePrice, minPrice, percentage);
        workingItems.add(item);
        persistedItems.add(item.clone());
        return Optional.of(item);
    }

    public Optional<MarketItem> getPersistedItem(String name) {
        for (MarketItem item : persistedItems) {
            if (item.getName().equals(name)) {
                return Optional.of(item);
            }
        }
        return Optional.empty();
    }

    public Optional<MarketItem> getWorkingItem(String itemName) {
        for (MarketItem item : workingItems) {
            if (item.getName().equals(itemName)) {
                return Optional.of(item);
            }
        }
        return Optional.empty();
    }

    public Optional<MarketItem> getPersistedItem(MarketItem item) {
        return getPersistedItem(item.getName());
    }

    public boolean removeItem(String name) {
        Optional<MarketItem> item = getPersistedItem(name);
        if (item.isEmpty()) return false;

        persistedItems.remove(item.get());
        workingItems.remove(item.get());
        return database.removeItem(name);
    }

    public boolean removeItem(MarketItem item) {
        return removeItem(item.getName());
    }

    public List<MarketItem> getPersistedItems() {
        return persistedItems;
    }
    public List<MarketItem> getWorkingItems() {
        return workingItems;
    }

    public MarketInitializer getInitializer() {
        return initializer;
    }

    public MarketDatabaseHandler getDatabaseHandler() {
        return databaseHandler;
    }

    public PriceHandler getPriceHandler() {
        return priceHandler;
    }
}

File: src/main/kotlin/me/wuzzyxy/dynamicmarket/market/MarketManager.kt
package me.wuzzyxy.dynamicmarket.market

import me.wuzzyxy.dynamicmarket.DynamicMarket
import me.wuzzyxy.dynamicmarket.database.Database
import me.wuzzyxy.dynamicmarket.items.MarketItem
import org.bukkit.Bukkit
import kotlin.math.max

class MarketManager(private val plugin: DynamicMarket, private val database: Database) {

    val workingItems: MutableList<MarketItem> = mutableListOf()
    val persistedItems: MutableList<MarketItem> = mutableListOf()

    val databaseHandler: MarketDatabaseHandler
    val priceHandler: PriceHandler
    val report: MarketReport
    val eventManager: MarketEventManager

    private var reportTask = NO_TASK

    init {
        val stored = database.getAllItems()
        if (stored == null) {
            plugin.logger.severe("Failed to load items from database")
        } else {
            persistedItems.addAll(stored)
            stored.mapTo(workingItems) { it.clone() }
        }

        databaseHandler = MarketDatabaseHandler(this, database, plugin)
        priceHandler = PriceHandler(plugin.pluginConfig.SELL_MULTIPLIER)
        report = MarketReport(plugin.pluginConfig.reportSettings(), { workingItems }, database, priceHandler)

        reconcileWithConfig(plugin.itemConfig, this, plugin.logger)
        startReportTask()

        // Constructed last: it reads plugin.eventConfig and the just-reconciled workingItems.
        eventManager = MarketEventManager(plugin, database, { workingItems })
    }

    /***
     * Pushes first so nothing in flight is lost, then rewrites the knobs in place rather
     * than rebuilding anything. The registered placeholder expansions hold references to
     * this manager, its price handler and its report — replace those objects and the
     * placeholders keep serving the old ones for the rest of the server's life.
     */
    fun reload() {
        databaseHandler.pushItems()

        priceHandler.sellMultiplier = plugin.pluginConfig.SELL_MULTIPLIER
        report.settings = plugin.pluginConfig.reportSettings()

        reconcileWithConfig(plugin.itemConfig, this, plugin.logger)

        databaseHandler.restartTasks()
        startReportTask()
        eventManager.reload()
    }

    private fun startReportTask() {
        Bukkit.getScheduler().cancelTask(reportTask)
        val ticks = max(1, plugin.pluginConfig.REPORT_REFRESH_SECONDS) * 20L
        reportTask = Bukkit.getServer().scheduler.scheduleSyncRepeatingTask(plugin, report::refresh, 220L, ticks)
    }

    /***
     * Returns the working item object
     */
    fun addItem(definition: MarketItem): MarketItem? {
        if (getPersistedItem(definition.name) != null) return null

        val item = database.addItem(definition) ?: return null
        workingItems.add(item)
        persistedItems.add(item.clone())
        return item
    }

    fun getPersistedItem(name: String): MarketItem? = persistedItems.firstOrNull { it.name == name }

    fun getPersistedItem(item: MarketItem): MarketItem? = getPersistedItem(item.name)

    fun getWorkingItem(itemName: String): MarketItem? = workingItems.firstOrNull { it.name == itemName }

    fun removeItem(name: String): Boolean {
        val item = getPersistedItem(name) ?: return false

        persistedItems.remove(item)
        workingItems.remove(item)
        return database.removeItem(name)
    }

    fun removeItem(item: MarketItem): Boolean = removeItem(item.name)

    private companion object {
        const val NO_TASK = -1
    }
}
